sympy/tensor/array/array_comprehension.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import print_function, division
+import functools
+from sympy.core.sympify import sympify
+from sympy.core.expr import Expr
+from sympy.core import Basic
+from sympy.core.compatibility import Iterable
+from sympy.tensor.array import MutableDenseNDimArray
+from sympy import  Symbol
+
+class ArrayComprehension(Basic):
+    """
+    Generate a list comprehension
+    """
+    def __new__(cls, expr, *bounds, **assumptions):
+        if any(len(l) != 3 or None for l in bounds):
+            raise ValueError('ArrayComprehension requires values lower and upper bound'
+                              ' for the expression')
+        cls.default_assumptions = assumptions
+        obj = Expr.__new__(cls, **assumptions)
+        obj.expr = expr
+        if cls._check_bounds_validity(bounds):
+            obj.bounds = bounds
+        arglist = [expr]
+        arglist.extend(obj.bounds)
+        obj._args = tuple(arglist)
+        return obj
+
+    @classmethod
+    def _check_bounds_validity(cls, bounds):
+        return True
+
+    def doit(self):
+        expr = self.expr
+        for index, value in enumerate(self.bounds):
+            var, inf, sup = map(sympify, value)
+            # Array will not ne expanded if there is a symbolic dimension
+            if Basic(inf, sup).atoms(Symbol):
+                return self
+        arr = self._expand_array()
+        return arr
+
+    # Add/replace a boudary of variable
+    def add_bound(self, bound):
```

Comment:
SymPy methods are immutable. No edits after class constructor, otherwise bad things may happen (i.e. problems with hashing, term rewriter, etc...).

Reply:
I didn't know that. Thank for telling me :-)